File: aautad_app/lib/models/news.dart
import '../utils/utils.dart'; // Import the fixEncoding function

class News {
  final String id;
  final String title;
  final String content;
  final String imageUrl;
  final DateTime createdAt;
  final DateTime updatedAt;

  News({
    required this.id,
    required this.title,
    required this.content,
    required this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory News.fromJson(Map<String, dynamic> json) {
    return News(
      id: json['id'],
      title: fixEncoding(json['title']), // Apply fixEncoding
      content: fixEncoding(json['content']), // Apply fixEncoding
      imageUrl: json['imageUrl'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'imageUrl': imageUrl,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

File: admin/src/server/api/routers/news.ts
import { createTRPCRouter, publicProcedure } from "../trpc";
import { z } from "zod";
import { instagramAPI } from "../../services/instagram-api";

export const newsSchema = z.object({
    title: z.string().min(1, "Title is required"),
    content: z.string().min(1, "Content is required"),
    type: z.enum(["main", "sports"]),
    imageUrl: z.string().url().optional(),
    type: z.string().min(1, "Type is required"),
    createdAt: z.date().default(() => new Date()),
    updatedAt: z.date().optional(),
});

export const newsRouter = createTRPCRouter({
    create: publicProcedure
        .input(newsSchema)
        .mutation(async ({ ctx, input }) => {
            const parsedData = newsSchema.parse(input);

            return ctx.db.news.create({
                data: {
                    ...parsedData,
                },
            });
        }),

    getAll: publicProcedure.query(async ({ ctx }) => {
        return ctx.db.news.findMany();
     }),

    edit: publicProcedure
        .input(
            z.object({
                id: z.string().optional(),
                newsSchema,
            })
        )
        .mutation(async ({ ctx, input }) => {
            const parsedData = newsSchema.parse(input.newsSchema);

            if(!input.id) return

            return ctx.db.news.update({
                where: {
                    id: input.id,
                },
                data: {
                    ...parsedData,
                },
            });
        }),

    delete: publicProcedure
        .input(
            z.object({
                id: z.string(),
            })
        )
        .mutation(async ({ ctx, input }) => {
            return ctx.db.news.delete({
                where: {
                    id: input.id,
                },
            });
        }),
        
    // New endpoint for publishing to Instagram
    publishToInstagram: publicProcedure
        .input(
            z.object({
                id: z.string().optional(),
                title: z.string(),
                content: z.string(),
                imageUrl: z.string().url(),
            })
        )
        .mutation(async ({ input }) => {
            try {
                if (!input.imageUrl) {
                    throw new Error("An image is required to publish to Instagram");
                }

                // Format the caption with title and content
                const caption = `${input.title}\n\n${input.content}`;

                // Call the Instagram API service to publish the post
                const result = await instagramAPI.publishPost({
                    caption,
                    imageUrl: input.imageUrl,
                });

                if (!result.success) {
                    throw new Error(result.error || "Failed to publish to Instagram");
                }

                return { 
                    success: true, 
                    instagramPostId: result.id,
                    message: "Successfully published to Instagram" 
                };
                
            } catch (error) {
                console.error("Instagram publishing error:", error);
                return { 
                    success: false, 
                    error: error instanceof Error ? error.message : "Unknown error occurred" 
                };
            }
        }),
});